
#### Overview of germline CNV data

```{r, eval=T, message=F}
library(karyoploteR)
library(vcfR)
## kind of don't need this?
##cnvIn=data$GATK4cnvVcf
##cnvIn=read.vcfR(cnvIn)
##cnvIn=cnvIn[-which(cnvIn@fix[ ,5])]
dGenes=data$gCNVAnnotSV
AnnotCNV=read.delim(dGenes)
```


```{r, eval=F}
library(ggplot2)
QS=extract.gt(cnvIn, "QS", as.numeric=T)
QA=extract.gt(cnvIn, "QA", as.numeric=T)
NP=extract.gt(cnvIn, "NP", as.numeric=T)
CN=extract.gt(cnvIn, "CN", as.numeric=T)
END=extract.info(cnvIn, "END")

df=data.frame(QS=QS, QA=QA, NP=NP, CN=CN)
colnames(df)=c("QS", "QA", "NP", "CN")

ggplot(df, aes(x=QA, y=QS, col=CN))+geom_point()
ggplot(df, aes(x=NP, y=QS/NP, col=CN))+geom_point()
```


```{r}
## Filter out gains vs del
gain=which(AnnotCNV$CN==3)
hetloss=which(AnnotCNV$CN==1)
homloss=which(AnnotCNV$CN==0)
```


```{r,fig.height=4, fig.width=5.5}
gainMat <- GRanges(seqnames=paste0("chr",AnnotCNV[gain, 2]), ranges=IRanges(start=AnnotCNV[gain, 3],
                      end=AnnotCNV[gain, 4]), path=AnnotCNV$Psource[gain],AF=ifelse(is.na(AnnotCNV$AF[gain]), 1, 1-AnnotCNV$AF[gain]))

gainMat$AFcol=brewer.pal(5, "Reds")[cut(gainMat$AF, c(0,  0.8, 0.9, 0.95, 1))]
gainMat$AFcol=ifelse(gainMat$path!="", "#67000d", gainMat$AFcol)

hetlossMat <-  GRanges(seqnames=paste0("chr",AnnotCNV[hetloss, 2]), ranges=IRanges(start=AnnotCNV[hetloss, 3], end=AnnotCNV[hetloss, 4]), path=AnnotCNV$Psource[hetloss],AF=ifelse(is.na(AnnotCNV$AF[hetloss]), 1, 1-AnnotCNV$AF[hetloss]))

hetlossMat$AFcol=brewer.pal(5, "Greens")[cut(hetlossMat$AF, c(0,  0.8, 0.9, 0.95, 1))]
hetlossMat$AFcol=ifelse(hetlossMat$path!="", "#006d2c", hetlossMat$AFcol)

homlossMat<- GRanges(seqnames=paste0("chr",AnnotCNV[homloss, 2]), ranges=IRanges(start=AnnotCNV[homloss, 3], end=AnnotCNV[homloss, 4]), path=AnnotCNV$Psource[homloss],AF=ifelse(is.na(AnnotCNV$AF[homloss]), 1, 1-AnnotCNV$AF[homloss]))

homlossMat$AFcol=brewer.pal(5, "Greens")[cut(homlossMat$AF, c(0,  0.8, 0.9, 0.95, 1))]
homlossMat$AFcol=ifelse(homlossMat$path!="", "#006d2c", homlossMat$AFcol)

##gnomadMat<-  GRanges(seqnames=paste0("chr",gnomadTestIn$Chromosome), ranges=IRanges(start=gnomadTestIn$Start,
##                      end=gnomadTestIn$End))
##gnomadMat2=gnomadMat[-which(duplicated(gnomadMat))]
##lx1=findOverlapPairs(gainMat, gnomadMat2)

kp <- plotKaryotype(chromosomes=c("autosomal"), genome="hg38", main = "Gain regions")
kpPlotRegions(kp, gainMat, col=gainMat$AFcol)
kp <- plotKaryotype(chromosomes=c("autosomal"), genome="hg38", main="Hom Loss regions")
kpPlotRegions(kp, homlossMat, col=homlossMat$AFcol)

```

#### ACMG class 4 or 5 regions

The following variants are level 4 or 5 based on ACMG guidelines

```{r, fig.height=4, fig.width=6}
filt2=which(AnnotCNV$ACMG_class%in%c(4,5))

ACMGMat <- GRanges(seqnames=paste0("chr",AnnotCNV[filt2, 2]), ranges=IRanges(start=AnnotCNV[filt2, 3],
                      end=AnnotCNV[filt2, 4]), path=AnnotCNV$Psource[filt2],AF=ifelse(is.na(AnnotCNV$AF[filt2]), 1, 1-AnnotCNV$AF[filt2]), SVtype=AnnotCNV$SV_type[filt2], Genes=AnnotCNV$Gene_name[filt2],phen=AnnotCNV$Pheno[filt2], ACMG=AnnotCNV$ACMG_class[filt2])
ACMGMat$AFcolR=brewer.pal(5, "Reds")[cut(ACMGMat$AF, c(0,  0.8, 0.9, 0.95, 1))]
ACMGMat$AFcolR=ifelse(ACMGMat$path!="", "#67000d", ACMGMat$AFcolR)
ACMGMat$AFcolB=brewer.pal(5, "Blues")[cut(ACMGMat$AF, c(0,  0.8, 0.9, 0.95, 1))]
ACMGMat$AFcolB=ifelse(ACMGMat$path!="", "#006d2c", ACMGMat$AFcolB)
ACMGMat$AFcolF=ifelse(ACMGMat$SVtype=="DUP", ACMGMat$AFcolR, ACMGMat$AFcolB)
                      
kp <- plotKaryotype(chromosomes=c("autosomal"), genome="hg38", main = "ACMG regions")
kpPlotRegions(kp, ACMGMat, col=ACMGMat$AFcolF)
```



```{r}
# ACMGGenes=strsplit(ACMGMat$Genes, ";")
# Genes2=sapply(ACMGGenes, function(x) x[which(x%in%GenesOfInterest)])
# Genes3=sapply(Genes2, function(x) paste(x, collapse=";"))
# df=data.frame(Chrom=seqnames(ACMGMat), Loc=ranges(ACMGMat), Type=ACMGMat$SVtype, GT=GT[filt2], path=ACMGMat$phen, AFmax=1-ACMGMat$AF, GenesofInterest=Genes3, Genes=ACMGMat$Genes, ACMG=ACMGMat$ACMG)

Scolnames=c("SV_chrom", "SV_start", "SV_end", "CN", "Pheno", "ACMG_class", "AF", 
            "GenesOfInterest","Cosmic", "Pathways","GTex", "Gene_name")
T1=AnnotCNV[filt2, Scolnames]
colnames(T1)[1:3]=c("CHROM", "START", "END")

DT::datatable(T1, rownames=F, filter = 'top', class='cell-border stripe',extensions="Buttons", options=list(dom="Bfrtip",scoller = TRUE, pageLength = 7,  buttons=c(I('colvis'),'csv', 'excel'), columnDefs = list(list(visible=TRUE))))

```

### Other Genes of Interest

The following plot shows regions of gains or losses, colored by:

* purple: in cosmic and in gene of interest list `r data$TreatmentKeywords`.
* blue: Gene of Interest
* red: Cosmic


```{r, fig.height=4}
lx1=which((AnnotCNV$GenesOfInterest!=""| AnnotCNV$Cosmic!="") & AnnotCNV$ACMG_class>=3 & (AnnotCNV$AF<=0.1 | is.na(AnnotCNV$AF)))

## Visualise the genes of interest?

temp1=AnnotCNV[lx1, c("SV_chrom", "SV_start" ,"CN", "GenesOfInterest", "Cosmic")]

tx2=grep(",", temp1$GenesOfInterest)

if (length(tx2)>0){
  Input2=strsplit(temp1$GenesOfInterest[tx2], ",") 
  lx3=sapply(Input2, length)
  tempMat=temp1[tx2, ]
  tempMat2=tempMat[rep(seq_len(nrow(tempMat)), times = lx3), ]
  tempMat2$GenesOfInterest=unlist(Input2)
  temp1=rbind(temp1[-tx2, ], tempMat2)
}

tx3=grep(",", temp1$Cosmic)

if (length(tx3)>0){
  Input2=strsplit(temp1$GenesOfInterest[tx3], ",") 
  lx3=sapply(Input2, length)
  tempMat=temp1[tx3, ]
  tempMat2=tempMat[rep(seq_len(nrow(tempMat)), times = lx3), ]
  tempMat2$GenesOfInterest=unlist(Input2)
  temp1=rbind(temp1[-tx3, ], tempMat2)
}

temp1=unique(temp1)
temp1$Col=ifelse((temp1$Cosmic!=""&temp1$GenesOfInterest!=""),"purple", ifelse(temp1$Cosmic!="", "red", "blue") )
temp1$Gene=ifelse(temp1$GenesOfInterest=="", temp1$Cosmic, temp1$GenesOfInterest)
temp1$Gene=gsub(paste0(" ",data$TreatmentKeywords), "", temp1$Gene)
temp1$Gene[which(temp1$Gene=="")]=NA

temp2=unique(temp1[ ,-c(4:5)])

temp2$SV_chrom=factor(temp2$SV_chrom, levels=c(1:22, "X", "Y"))
temp2=temp2[order(temp2$SV_chrom,temp2$SV_start), ]

tempx=by(temp2$CN, temp2$Gene, mean)
tempx2=stack(round(tempx))

temp2=temp2[-which(duplicated(temp2[ ,c(4:5)])), ]
temp2$CN=factor(tempx2$values[ match(temp2$Gene, tempx2$ind)])
temp2$Gene=factor(temp2$Gene, levels=temp2$Gene)

colB=temp2$Col

ggplot(temp2, aes(x=Gene, y=1, fill=CN))+geom_tile()+  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())+scale_fill_manual(values = c("#3182bd", "#9ecae1", "#de2d26"))+expand_limits(y = 0.25)+geom_text(aes(colour = colB, label = Gene), y = 0.35, angle=90)+facet_grid(~SV_chrom, space="free", scales="free")+scale_colour_manual(values = c("blue", "purple", "red"), guide = F)
```


```{r}
Scolnames=c("SV_chrom", "SV_start", "SV_end", "CN", "Pheno", "ACMG_class", "AF", 
            "GenesOfInterest","Cosmic", "Pathways","GTex", "Gene_name")
T1=AnnotCNV[lx1, Scolnames]
colnames(T1)[1:3]=c("CHROM", "START", "END")

DT::datatable(T1, rownames=F, filter = 'top', class='cell-border stripe',extensions="Buttons", options=list(dom="Bfrtip",scoller = TRUE, pageLength = 5,  buttons=c(I('colvis'),'csv', 'excel'), columnDefs = list(list(visible=FALSE, targets=c(4)))))

```
