
#### Overview of germline CNV data

Genes which are Cancer-COSMIC related, or have relevance to the pathway of interest (`r `data$TreatmentKeywords`)

* <span style="color: red;"> Gains or insertions (top) </span>
* <span style="color: blue;"> Deletions (Bottom) </span>

```{r, eval=T, message=F}
# divide the annotation outputs
T1=AnnotCNV$full
T2=AnnotCNV$split
T3=AnnotCNV$acmg
```

```{r,fig.height=6, message=F}
# split the gains and losses within the full Annotation for karyotype plotter
gain=which(T1$CN>2)
loss=which(T1$CN<2)

## need to debug this section
gainMat=tryCatch({CreateGRangesData(T1[gain, ], "Reds",F)}, error=function(e){GRanges()})
lossMat=tryCatch({CreateGRangesData(T1[loss, ], "Blues",F)}, error=function(e){ GRanges()})


#gainMat <- GRanges(seqnames=paste0("chr",T1[gain, "CHROM"]), ranges=IRanges(start=T1[gain, "START"],
#                      end=T1[gain, "END"]), path=T1$PathogenicSource[gain], AF=ifelse(is.na(T1$Pop.Freq[gain]), 1, 1-T1$Pop.Freq[gain]))
#lossMat<-GRanges(seqnames=paste0("chr",T1[loss, "CHROM"]), ranges=IRanges(start=T1[loss, "START"], end=T1[loss, "END"]), path=T1$PathogenicSource[loss],
#                 AF=ifelse(is.na(T1$Pop.Freq[loss]), 1, 1-T1$Pop.Freq[loss]))

#gainMat$AFcol=brewer.pal(5, "Reds")[cut(gainMat$AF, c(0,  0.8, 0.9, 0.95, 1))]
#gainMat$AFcol=ifelse(gainMat$path!="", "#67000d", gainMat$AFcol)

#lossMat$AFcol=brewer.pal(5, "Blues")[cut(lossMat$AF, c(0,  0.8, 0.9, 0.95, 1))]
#lossMat$AFcol=ifelse(lossMat$path!="", "#006d2c", lossMat$AFcol)

# find the location of all the genes of interest

if (length(T2$Gene_name)>1){
  genes=genesAll[na.omit(match(T2$Gene_name, genesAll$hgnc_symbol))]
}

kp <- plotKaryotype(chromosomes=c("autosomal"), genome="hg38",plot.type=2)
ax=tryCatch({kpPlotRegions(kp, gainMat, col=gainMat$AFcol,data.panel = 1)}, error=function(e){NULL})
ax=tryCatch({kpPlotRegions(kp, lossMat, col=lossMat$AFcol,data.panel = 2)}, error=function(e){NULL})
ax=tryCatch({kpPlotMarkers(kp, data=genes, labels=genes$hgnc_symbol, text.orientation = "horizontal", r1=0.75, cex=0.5, adjust.label.position = FALSE)}, error=function(e){NULL})

```

#### Genes/Pathways of Interest

The following plot shows regions of gains or losses which is multi region spanning (breakpoints expected to cross entire transcript, an exon or beyond a given intron), colored by:

* <span style="color: red;"> Cosmic Cancer Gene (Tier 1 or 2 oncogene or tumor suppressor gene) </span>
* <span style="color: blue;"> Impacts the pathway of interest: `r data$TreatmentKeywords`, either through MSigDB annotation or user-defined gene list </span>
* <span style="color: purple;"> Gene is both in Cosmic List and is a gene of interest list in `r data$TreatmentKeywords` pathway </span>

```{r, fig.height=3.5}
## Visualise the genes of interest
temp1=T2[which(T2$MultiSpan), c("CHROM", "START", "END","Gene_name", "Cosmic", "PathwayRelated","CN", "Location")]
temp1$Col=ifelse((temp1$Cosmic!=""&temp1$PathwayRelated!=""),"purple", ifelse(temp1$Cosmic!="", "red", "blue") )
temp1$CN=factor(temp1$CN, levels=c("0", "1","2", "3", "4"))
temp1$CN[1]=0
fillvals=list("0"="#3182bd", "1"="#9ecae1", "2"="white", "3"="#fc9272","4"="#de2d26")

ggplot(temp1, aes(x=Gene_name, y=1, fill=CN))+geom_tile()+theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())+scale_fill_manual(values = fillvals)+expand_limits(y = 0.25)+geom_text(aes(colour = Col, label = Gene_name), y = 0.35, angle=90)+facet_grid(~CHROM, space="free", scales="free")+scale_colour_manual(values = c("blue", "purple", "red"), guide = "none")
```

**Multi-gene variants**

```{r}
idx=which(T1$GenesOfInterest!=""|T1$Cosmic!="")
MGexist=ifelse(length(idx)>0, T, F)
```

`r ifelse(MGexist==F, print('There are no CNVs spanning multiple genes'), print('The CNVs spanning multiple genes are shown below:'))`

```{r, eval=MGexist}
# read in the table here
DT::datatable(T1[idx, ], rownames=F, filter = 'top', class='cell-border stripe',extensions="Buttons", options=list(dom="Bfrtip",scoller = TRUE, pageLength = 5,  buttons=c(I('colvis'),'csv', 'excel'), columnDefs = list(list(visible=FALSE, targets=c(4,5,10,12,14,22,23)))))

```

**Gene level results**

The following fields are gene level annotations from AnnotSV: please see this documentation for more information: https://lbgi.fr/AnnotSV/Documentation/README.AnnotSV_latest.pdf

* GenCC Haploinsufficiency,Triplosensitivity: [0-3] 3 max score for dosage pathogenicity. Scores of 40/30 evidence for no dosage sensitivity and autosomal recessive
* OMIM: disease annotation from OMIM and morbidity from loss
* pLI: intolerance to protein truncating variants from gnomad or ExAC [0 -1]: closer to 1 is predicted intolerance


```{r}
# read in the table here
DT::datatable(T2, rownames=F, filter = 'top', class='cell-border stripe',extensions="Buttons", options=list(dom="Bfrtip",scoller = TRUE, pageLength = 5,  buttons=c(I('colvis'),'csv', 'excel'), columnDefs = list(list(visible=FALSE, targets=c(6,8,12)))))

```


#### ACMG class 4 or 5 regions

The following variants are level 4 or 5 based on AnnotSV ACMG prediction. Review the genes and population frequencies to assess whether they are relevant or not:

```{r, fig.height=6}
ACMGMat <- GRanges(seqnames=paste0("chr",T3$CHROM), ranges=IRanges(start=T3$START,end=T3$END),AF=ifelse(is.na(T3$Pop.Freq), 1, 1-T3$Pop.Freq), path=T3$PathogenicSource,Genes=T3$Gene_name,phen=T3$Pheno, gainloss=ifelse(T3$CN>2, "gain", "loss"))

#calculate gains
ACMGMat$AFcolR=brewer.pal(6, "Reds")[cut(ACMGMat$AF, c(0,0.5,  0.8, 0.9, 0.95, 1))]
ACMGMat$AFcolR=ifelse(ACMGMat$path!="", "#67000d", ACMGMat$AFcolR)
#calculate loss
ACMGMat$AFcolB=brewer.pal(6, "Blues")[cut(ACMGMat$AF, c(0,0.5,  0.8, 0.9, 0.95, 1))]
ACMGMat$AFcolB=ifelse(ACMGMat$path!="", "#006d2c", ACMGMat$AFcolB)

if (length(T3$Gene_name)>1){
  genes=genesAll[na.omit(match(T3$Gene_name, genesAll$hgnc_symbol))]
}
                      
kp <- plotKaryotype(chromosomes=c("autosomal"), genome="hg38", main = "ACMG regions", plot.type=2)
ax=tryCatch({kpPlotRegions(kp, ACMGMat[which(ACMGMat$gainloss=="gain")], col=ACMGMat$AFcolR,data.panel = 1)},  error=function(e){NULL})
ax=tryCatch({kpPlotRegions(kp, ACMGMat[which(ACMGMat$gainloss=="del")], col=ACMGMat$AFcolB,data.panel = 2)}, error=function(e){NULL})
ax=tryCatch({kpPlotMarkers(kp, data=genes, labels=genes$hgnc_symbol, text.orientation = "horizontal", r1=0.75, cex=0.5, adjust.label.position = FALSE)}, error=function(e){NULL})

```



```{r}
#read in the table here
DT::datatable(T3, rownames=F, filter = 'top', class='cell-border stripe',extensions="Buttons", options=list(dom="Bfrtip",scoller = TRUE, pageLength = 7,  buttons=c(I('colvis'),'csv', 'excel')))
```

