
Here we report any genes which could be impacted by a combination of SNVs, CNV and/or SVs, below is an overlap of the types of variation found in genes of interest:

```{r all-find-intersect, fig.height=5, fig.width=5}
## combine all SNVs together
STerms=c("SYMBOL", "rsID", "CHROM", "POS", "HGVSp", "Consequence", "gnomAD.AF", "ClinVar.Sig","ClinVar.Disease", "ProteinDomain", "CancerGeneCensus.Tier","CancerMutationCensus.Tier")
allSNV=rbind(D2[, STerms], D3[ ,STerms], D4[ ,STerms], D5[ ,STerms], D6[ ,STerms])
allSNV$class=c(rep("Cosmic", nrow(D2)), rep("Pathway", nrow(D3)),
                rep("Drug", nrow(D4)), rep("VUS", nrow(D5)), rep("ACMG", nrow(D6)))
allSNV$Genomic.Pos=paste(allSNV$CHROM, allSNV$POS, allSNV$HGVSp)

# SV terms - use S1
# allSV=rbind(PGenesSVfilt, CGenesSVfilt)
# allSV$class=c(rep("Pathway", nrow(PGenesSVfilt)),rep("Cosmic", nrow(CGenesSVfilt)))
# 
# # merge together the information
# t1=paste(allSNV$CHROM, allSNV$POS, allSNV$HGVSp)
# t2=paste0("SNV ", allSNV$Consequence)
# 
# 
# ## centralise the names of the samples
# allSNVb=data.frame(SYMBOL=allSNV$SYMBOL, Location=allSNV$ProteinDomain, PopFreq=allSNV$gnomAD.AF, VariantType=t2,Pathogenicity=allSNV$ClinVar.Sig, Disease=allSNV$ClinVar.Disease, VariantCategory=allSNV$class)
# 
# if (nrow(allSNVb)>0){
#   allSNVb$Genomic.Pos=t1
# }
# 
# 
# t3=ifelse(allSV$Haploinsufficiency>1 & allSV$Haploinsufficiency<=3, "Haploinsufficient", 
#           ifelse(allSV$Triplosensitivity>1 & allSV$Triplosensitivity<=3, "Triplosensitive", 
#                  ifelse(allSV$PathogenicSource!="", "Possibly pathogenic", 
#                  ifelse(allSV$BenignSource!="", "Possibly benign", "unknown"))))
# 
# # bug in binding when there are 0 rows
# allSVb=data.frame(SYMBOL=allSV$GENE, Genomic.Pos= allSV$SV_ID, Location=allSV$Location, PopFreq=allSV$`Pop Freq`, Pathogenicity=t3,   Disease=allSV$OMIM, VariantCategory=allSV$class)
# 
# if (nrow(allSVb)>0){
#  allSVb$VariantType=paste0("SV ", allSV$TYPE)
# }
# 
# # put in a check to make sure dimension is big enough
# if (nrow(T1b)>0){
# t4=paste0(T1b$CHROM, ":", T1b$START,"-", T1b$END)
# t5=ifelse(T1b$Haploinsufficiency>1 & T1b$Haploinsufficiency<=3, "Haploinsufficient", 
#           ifelse(T1b$Triplosensitivity>1 & T1b$Triplosensitivity<=3, "Triplosensitive", 
#                  ifelse(T1b$PathogenicSource!="", "Possibly pathogenic", 
#                  ifelse(T1b$BenignSource!="", "Possibly benign", "unknown"))))
# 
# allCNVb=data.frame(SYMBOL=T1b$Gene_name, Genomic.Pos=t4, Location=T1b$Location, PopFreq=T1b$Pop.Freq, VariantType=paste0("CN ", T1b$CN),Pathogenicity=t5, Disease=T1b$OMIM, VariantCategory=ifelse(T1b$Cosmic=="", "Pathway", "Cosmic"))
# }else{
#   print('no variants')
#   CNames=c("SYMBOL", "Genomic.Pos", "Location", "PopFreq", 
#                       "VariantType","Pathogenicity", "Disease", "VariantCategory")
#   allCNVb=data.frame(matrix(ncol=length(CNames), nrow=0))
#   colnames(allCNVb)=CNames
# }

#venn(list(SNV=unique(allSNVb$SYMBOL), SV=unique(allSVb$SYMBOL), CNV=unique(allCNVb$SYMBOL)))
venn(list(SNV=unique(allSNV$SYMBOL), SV=unique(S2$GENE), CNV=unique(T2$GENE)))
```

**Common genes**

Below are common genes between different variant types:

* SNVs are coloured according to consequence
* SVs coloured by SV type
* CNVs according to copy number

```{r all-plot-intersect, fig.height=5}
a1=intersect(allSNV$SYMBOL,S2$GENE)
a2=intersect(allSNV$SYMBOL,T2$GENE)
a3=intersect(S2$GENE,S3$GENE)
a4=unique(c(a1, a2))
contRun=ifelse(length(a4)>1, T, F)
```

`r ifelse(contRun,print('') ,print('There are no common variants with SNVs/CNV/SVs'))`

```{r, eval=contRun}
VariantTable=data.frame(gene=a4, SNV=NA, SV=NA, CNV=NA)
VariantTable$SNV=allSNV$Consequence[match(a4, allSNV$SYMBOL)]
VariantTable$SV=S2$TYPE[match(a4, S2$GENE)]
VariantTable$CNV=T2$CN[match(a4, T2$GENE)]

VariantTable

mVT=melt(VariantTable, measure.vars=c("SNV", "SV", "CNV"))

ggplot(mVT, aes(x=gene, y=variable, fill=value))+geom_tile()+theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
``` 
 
**Summary Table of genes**
 
`r ifelse(contRun,print('') ,print('There are no common variants with SNVs/CNV/SVs'))`
 
 
```{r all-print-variants, eval=contRun}
nSNV=allSNV[which(allSNV$SYMBOL%in%a4), c("SYMBOL", "CHROM", "POS", "HGVSp", "Consequence", "ProteinDomain", "class")]
nCV=T2[which(T2$GENE%in%a4), c("GENE", "START", "END", "CN", "Location")]
colnames(nCV)=paste(colnames(nCV), ".CNV", sep="")
colnames(nCV)[1]="SYMBOL"
nSV=S2[which(S2$GENE%in%a4), c("GENE", "START", "END", "TYPE", "Location")]
colnames(nSV)=paste(colnames(nSV), ".SV", sep="")
colnames(nSV)[1]="SYMBOL"

  
comb1<-tryCatch({left_join(nSNV, nCV, by = c("SYMBOL"))}, error=function(e){nSNV})
comb2<-tryCatch({left_join(comb1, nSV, by = c("SYMBOL"))}, error=function(e){comb1}) 

DT::datatable(comb2,rownames=F, filter = 'top', class='cell-border stripe',extensions="Buttons", options=list(dom="Bfrtip",scoller = TRUE, pageLength = 5,  buttons=c(I('colvis'),'csv', 'excel')))
```

