
#### Overview of germline SV data

Summary of the number of variants:

* ACMG score `r paramfile$AnnotSVACMGcutoff`
* Variant Allele Frequency above `r paramfile$SVVAF`
* Minimum number of supporting split reads `r paramfile$SVSRfilter`
* Minimum number of supporting paired reads `r paramfile$SVPRfilter`

```{r, message=FALSE}
suppressMessages(library(karyoploteR))
## Filter out gains vs del
gain=which(AnnotSV$SV_type=="DUP")
homloss=which(AnnotSV$SV_type=="DEL")
ins=which(AnnotSV$SV_type=="INS")
inv=which(AnnotSV$SV_type=="INV")
bnd=which(AnnotSV$SV_type=="BND")

VariantSummary=c("Number of DUP/GAIN SVs ", "Number of DEL SVs ",
             "Number of INS SVs ", "Number of INV SVs ",
             "Number of BND SVs ")

Nx=data.frame(Ngain=length(gain), Ndel=length(homloss), Nins=length(ins), Ninv=length(inv), Nbnd=length(bnd))
Nx= cbind(VariantSummary, t(Nx))

kable(t(Nx), rownames = F)
```
The following variants are color coded:

* <span style="color: red;"> Gains or insertions (top) </span>
* <span style="color: blue;"> Deletions (bottom) </span>
* <span style="color: green;"> Inversions and other translocations (top) </span>

Cosmic related genes or genes appearing in user defined list of interest are indicated

```{r, fig.height=5, message=FALSE}
CreateGRangesData=function(AnnotT, Col, Ncheck){
  gMat <- GRanges(seqnames=paste0("chr",AnnotT[ ,2]),ranges=IRanges(start=AnnotT[, 3], end=AnnotT[, 4]),names=AnnotT$AnnotSV_ID, path=AnnotT$Psource, AF=AnnotT$MeanVAF)
  gMat$AFcol=brewer.pal(6, Col)[cut(gMat$AF, c(0,0.25, 0.5, 0.8, 0.9,  1))]
  tx1=as.matrix(dist(start(gMat)))
  tx2=as.matrix(dist(end(gMat)))
  ReviewThese=which(tx1<Ncheck & tx2<Ncheck, arr.ind=T)
  ReviewThese=ReviewThese[which(ReviewThese[ ,1]<ReviewThese[ ,2]), ]
  if (length(ReviewThese)>0 ){
    Tx2=which(ReviewThese[ ,1]!=ReviewThese[ ,2])
    tx3=ReviewThese[ which(ReviewThese[ ,1]!=ReviewThese[ ,2]), ]
    CheckChr=which(seqnames(gMat)[tx3[ ,1]]==seqnames(gMat)[tx3[ ,2]])
    tx4=tx3[CheckChr, ]
  ## Merge the samples
    for (i in 1:nrow(tx4)){
      end(gMat[tx4[i, 1]])=end(gMat[tx4[i, 2]]) 
      gMat$AF[tx4[i, 1]]=mean(gMat$AF[tx4[i, ]])
    }
  gMat=gMat[-tx4[,2 ]]
}
  return(gMat)  
}
gainMat2=tryCatch({CreateGRangesData(AnnotSV[c(gain, ins), ], "Reds", 500)}, error=function(e){
  GRanges()})
lossMat2=tryCatch({CreateGRangesData(AnnotSV[homloss, ], "Blues", 500)}, error=function(e){
  GRanges()})
invMat2=tryCatch({CreateGRangesData(AnnotSV[c(inv, bnd)], "Green", 500)}, error=function(e){
  GRanges(names=NULL, path=NULL, AF=NULL)})

#lx1=which((AnnotSV$GenesOfInterest!=""| AnnotSV$Cosmic!="") & AnnotSV$ACMG_class>=3 & (AnnotSV$AF<=0.1 | is.na(AnnotSV$AF)) & AnnotSV$MeanVAF>0.25)

UniqueGenesA=unique(unlist(strsplit(AnnotSV$GenesOfInterest, " ")))
UniqueGenesA=unique(unlist(strsplit(UniqueGenesA, ",")))
UniqueGenesB=unique(unlist(strsplit(AnnotSV$Cosmic, ", ")))
UniqueGenesC=AnnotSV$Gene_name[grep(data$TreatmentKeywords, AnnotSV$GenesOfInterest)]

Tx3=na.omit(unique(c(UniqueGenesA, UniqueGenesB, UniqueGenesC)))
Tx3=setdiff(Tx3, data$TreatmentKeywords)

# ensembl <- useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl")
# genes <- toGRanges(getBM(attributes=c('chromosome_name', 'start_position', 'end_position', 'hgnc_symbol'),
#                filters = 'hgnc_symbol', values =Tx3, mart = ensembl))
# seqlevelsStyle(genes) <- "UCSC"
genes=genesAll[na.omit(match(Tx3, genesAll$hgnc_symbol))]

kp <- plotKaryotype(chromosomes=c("autosomal"), genome="hg38", plot.type=2)
ax=tryCatch({kpPlotRegions(kp, gainMat2, col=gainMat2$AFcol,data.panel = 1)}, error=function(e){NULL})
ax=tryCatch({kpPlotRegions(kp, lossMat2, col=lossMat2$AFcol,data.panel = 2)}, error=function(e){NULL})
ax=tryCatch({kpPlotRegions(kp, invMat2, col=invMat2$AFcol,data.panel = 3)}, error=function(e){NULL})
ax=tryCatch({kpPlotMarkers(kp, data=genes, labels=genes$hgnc_symbol, text.orientation = "horizontal",
              r1=0.75, cex=0.5, adjust.label.position = FALSE)}, error=function(e){NULL})
```


```{r, fig.height=3.5, eval=F}
lx1=which((AnnotSV$GenesOfInterest!=""| AnnotSV$Cosmic!="") & AnnotSV$ACMG_class>=paramfile$AnnotSVACMGcutoff & (AnnotSV$AF<=0.1 | is.na(AnnotSV$AF)) & AnnotSV$MeanVAF>0.25)

## Visualise the genes of interest?
temp1=AnnotSV[lx1, c("SV_chrom", "SV_start", "SV_end", "SV_type", "GenesOfInterest", "Cosmic")]

tx2=grep(",", temp1$GenesOfInterest)

if (length(tx2)>0){
  Input2=strsplit(temp1$GenesOfInterest[tx2], ",") 
  lx3=sapply(Input2, length)
  tempMat=temp1[tx2, ]
  tempMat2=tempMat[rep(seq_len(nrow(tempMat)), times = lx3), ]
  tempMat2$GenesOfInterest=unlist(Input2)
  temp1=rbind(temp1[-tx2, ], tempMat2)
}

tx3=grep(",", temp1$Cosmic)

if (length(tx3)>0){
  Input2=strsplit(temp1$GenesOfInterest[tx3], ",") 
  lx3=sapply(Input2, length)
  tempMat=temp1[tx3, ]
  tempMat2=tempMat[rep(seq_len(nrow(tempMat)), times = lx3), ]
  tempMat2$GenesOfInterest=unlist(Input2)
  temp1=rbind(temp1[-tx3, ], tempMat2)
}

temp1=unique(temp1)
temp1$GenesOfInterest[which(is.na(temp1$GenesOfInterest))]=""
temp1$Col=ifelse((temp1$Cosmic!=""&temp1$GenesOfInterest!=""),"purple", ifelse(temp1$Cosmic!="",  "red", "blue"))

temp1$Gene=ifelse((temp1$GenesOfInterest==""| is.na(temp1$GenesOfInterest)), temp1$Cosmic, temp1$GenesOfInterest)
temp1$Gene=gsub(data$TreatmentKeywords, "", temp1$Gene)
temp1$Gene=gsub(" ", "", temp1$Gene)
#
#
temp1$Gene[which(temp1$Gene=="")]=NA

temp2=unique(temp1[ ,-c(5:6)])

temp2$SV_chrom=factor(temp2$SV_chrom, levels=c(1:22, "X", "Y"))
temp2=temp2[order(temp2$SV_chrom,temp2$SV_start), ]

#tempx=by(temp2$CN, temp2$Gene, mean)
#tempx2=stack(round(tempx))

#temp2=temp2[-which(duplicated(temp2[ ,c(4:5)])), ]
# temp2$CN=(tempx2$values[ match(temp2$Gene, tempx2$ind)])
# temp2$CN[which(temp2$CN==2)]=1
# temp2$CN=factor(temp2$CN)
temp2$Gene=factor(temp2$Gene, levels=temp2$Gene)

colB=temp2$Col

ggplot(temp2, aes(x=Gene, y=1, fill=SV_type))+geom_tile()+  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())+scale_fill_manual(values = c("#3182bd", "#9ecae1", "#de2d26"))+expand_limits(y = 0.25)+geom_text(aes(colour = colB, label = Gene), y = 0.35, angle=90)+facet_grid(~SV_chrom, space="free", scales="free")+scale_colour_manual(values = c("blue", "purple", "red"), guide = F)
```

**Multi-gene spanning SVs**


```{r}
lx1=which((AnnotSV$GenesOfInterest!=""| AnnotSV$Cosmic!="") & AnnotSV$ACMG_class>=paramfile$AnnotSVACMGcutoff & (AnnotSV$AF<=paramfile$OverallGnomad | is.na(AnnotSV$AF)) & AnnotSV$MeanVAF>=paramfile$SVVAF)
ScolnamesTable=cbind(c("SV_chrom", "SV_start", "SV_end","SV_type", "SV_length", "ACMG_class", "AF", "GT","MeanVAF","SRVAF", "GenesOfInterest","Cosmic", "Pathways","GTex", "Gene_name",  "Pheno"), c("CHROM", "START", "END", "TYPE", "SV_length", "ACMG", "Pop Freq", "GT", "VAF", "SR-VAF", "GenesOfInterest", "Cosmic", "Pathways", "GTex", "Gene Names", "CLinVar Phenotypes"))

midx=match(ScolnamesTable[ ,1], colnames(AnnotSV))

T1=AnnotSV[lx1, na.omit(midx)]
colnames(T1)=ScolnamesTable[which(!is.na(midx)) ,2]

DT::datatable(T1, rownames=F, filter = 'top', class='cell-border stripe',extensions="Buttons", options=list(dom="Bfrtip",scoller = TRUE, pageLength = 5,  buttons=c(I('colvis'),'csv', 'excel'), columnDefs = list(list(visible=FALSE, targets=c(12:14)))))
```

#### Gene level summary

The following fields are gene level annotations from AnnotSV: please see this documentation for more information: https://lbgi.fr/AnnotSV/Documentation/README.AnnotSV_latest.pdf

* GenCC Haploinsufficiency,Triplosensitivity: [0-3] 3 max score for dosage pathogenicity. Scores of 40/30 evidence for no dosage sensitivity and autosomal recessive
* OMIM: disease annotation from OMIM and morbidity from loss
* pLI: intolerance to protein truncating variants from gnomad or ExAC [0 -1]: closer to 1 is predicted intolerance

```{r,message=F}
dGenes=data$gSVsplitAnnotSV
AnnotSVGenes=read.delim(dGenes)
#AnnotSVGenes=read.delim("/Users/anntri/gitLibs/WDLPipelines/cromwell-executions/ClinicalReport/7ca9d50f-4214-4abb-9b14-f72cf3ea114f/call-CreateClinical/execution/ER095_MEL3N.SV.split.formated.tsv")
AnnotSVGenes$OMIM_morb=ifelse(AnnotSVGenes$OMIM_morbid=="yes", "morbid", ifelse(AnnotSVGenes$OMIM_morbid_candidate=="yes", "candidate", "no"))
#AnnotSV=AnnotSV2[which(AnnotSV2$ACMG_class%in%c(3,4,5) & AnnotSV2$MeanVAF>0.25), ]
```

**COSMIC related genes**

```{r}
CGenes=which(AnnotSVGenes$Cosmic!="" & AnnotSVGenes$MeanVAF>=paramfile$SVVAF)
CGenesSV=AnnotSVGenes[CGenes, ]
Scolnames=c("Gene_name","AnnotSV_ID", "SV_type", "SV_length", "ACMG_class", "Location", "AF", "HI", "TS","OMIM_phenotype", "OMIM_morb",
            "RE_gene", "Dist_nearest_SS", "TAD_coordinate",
             "Bsource", "Psource", "GnomAD_pLI", "ExAC_pLI", "SRVAF",
            "MeanVAF", "Depth")
CGenesSVfilt=CGenesSV[ ,Scolnames]

colnames(CGenesSVfilt)=c("GENE","SV_ID", "TYPE", "SV_length", "ACMG", "Location", "Pop Freq", "Haploinsufficiency", "Triplosensitivity", "OMIM", "OMIM_morbid", "Reg.Element.Genes", "Nearest.Splice.Site.bp", "TAD", "BenignSource", "PathogenicSource", "gnomad.pLI", "exac.pLI", "SplitReadVAF", "MeanVAF", "Depth")

DT::datatable(CGenesSVfilt, rownames=F, filter = 'top', class='cell-border stripe',extensions="Buttons", options=list(dom="Bfrtip",scoller = TRUE, pageLength = 7,  buttons=c(I('colvis'),'csv', 'excel'), columnDefs = list(list(visible=FALSE, targets=c(9, 11, 12, 13,14, 17:20)))))

```

**Pathway related genes**

```{r}
PGenes=which(AnnotSVGenes$GenesOfInterest!="" & AnnotSVGenes$MeanVAF>=paramfile$SVVAF)
PGenesSV=AnnotSVGenes[PGenes, ]
Scolnames=c("Gene_name","AnnotSV_ID", "SV_type", "SV_length", "ACMG_class", "Location", "AF", "HI", "TS","OMIM_phenotype", "OMIM_morb",
            "RE_gene", "Dist_nearest_SS", "TAD_coordinate",
             "Bsource", "Psource", "GnomAD_pLI", "ExAC_pLI","SRVAF",
            "MeanVAF", "Depth")
PGenesSVfilt=PGenesSV[ ,Scolnames]

colnames(PGenesSVfilt)=c("GENE","SV_ID", "TYPE", "SV_length", "ACMG", "Location", "Pop Freq", "Haploinsufficiency", "Triplosensitivity", "OMIM", "OMIM_morbid", "Reg.Element.Genes", "Nearest.Splice.Site.bp", "TAD", "BenignSource", "PathogenicSource", "gnomad.pLI", "exac.pLI", "SplitReadVAF", "MeanVAF", "Depth")

DT::datatable(PGenesSVfilt, rownames=F, filter = 'top', class='cell-border stripe',extensions="Buttons", options=list(dom="Bfrtip",scoller = TRUE, pageLength = 7,  buttons=c(I('colvis'),'csv', 'excel'), columnDefs = list(list(visible=FALSE, targets=c(9, 11, 12, 13,14, 17:20)))))
              

```


```{r, eval=F}
## Bed annotation for annotSV
bed2=AnnotSV[ c("SV_chrom", "SV_start", "SV_end",  "SV_length","SV_type", "MeanVAF","chr2", "loc2", "Gene_name")]
bed2[ ,1]=paste("chr", bed2[ ,1], sep="")


bed=A1[ ,c("CONTIG","START","END", "MEAN_LOG2_COPY_RATIO")]
colnames(bed)=c("chr","start","end", "value1")
bed[ ,1]=paste("chr", bed[,1], sep="")

bedt=do.call(rbind, Xout[2:4])
bed3=cbind(bedt[ ,c("Chromosome", "Start_Position", "End_Position")], value=substr(rownames(bedt), 1, 1))
bed3$value=gsub("F", 1,bed3$value)
bed3$value=gsub("P", 2,bed3$value)
bed3$value=gsub("V", 3,bed3$value)

bed3[ ,1]=paste("chr", bed3[,1], sep="")
bed3$value=as.numeric(bed3$value)

colnames(bed3)=c("chr","start","end","value1")
## include the bed file for actionable mutations?
## insert gene names of actionable mutations, or those which have a CNV
bed4=bed3[ which(bed3$value!=3), c(1:4) ]

circos.par("track.height" = 0.08)
circos.initializeWithIdeogram(species="hg19") # plot the ideogram


circos.genomicTrack(bed3, stack=TRUE, panel.fun = function(region,value,...) {
    circos.genomicPoints(region, value, cex = 0.75, pch = 20,col=ifelse(value==1,"#bd0026", ifelse(value==2, "#f03b20", "#fed976")), ...)
})

circos.genomicTrack(bed, panel.fun = function(region, value, ...) {
    circos.genomicPoints(region, value, cex = 0.75, pch = 20,col=ifelse(value>0.6, "red", ifelse(value<(-1), "blue", "grey")), ...)
})

#circos.genomicTrack(bed, panel.fun = function(region, value, ...) {    circos.genomicPoints(region, value, cex = 0.5, pch = 16,col=ifelse(value>0.6, "red", ifelse(value<(-1), "blue", "grey")), ...)})

if (!is.null(data$Manta)){

typeE=c("DEL","DUP","INS")
colE=c("blue","red","green")
for (i in 1:3) {
        bed_list=bed2[which(bed2[,6]==typeE[i]),c(1:4)]
        if (dim(bed_list)[1]>0){
        circos.genomicTrack(bed_list, track.height=0.05,
    panel.fun = function(region, value, ...) {
        circos.genomicRect(region, value, ytop.column = 1, ybottom = 0, 
            col = colE[i], ...)
        circos.lines(CELL_META$cell.xlim, c(0, 0), lty = 2, col = "#00000040")
    })
    } else {
        circos.genomicTrackPlotRegion(bed_list,track.height=0.05,stack=TRUE, panel.fun = function(region, value, ...) {circos.genomicRect(region, value, cex = 0.5, pch = 16, col = colE[i], ...)})
    }

}

## plot the last part of this
bed2b=bed2[which(bed2[ ,6]=="BND"), c(1,2, 7:8)]
if (dim(bed2b)[1]>0){
for (i in 1:dim(bed2b)[1]) {
        circos.link(bed2b[i,1],as.numeric(bed2b[i,2]),bed2b[i,3],as.numeric(bed2b[i,4]))
}}
}

title("Somatic calls (SNV - CNV - SV)")
legend(0.7,1.05,legend=c("SNV", "CNV","SV-DEL","SV-DUP","SV-INS"),col=c("orange","grey","blue","red","green"),pch=c(16,16,15,15,15),cex=0.75,title="Tracks:",bty='n')
legend(0.7,0.8,legend="SV-TRANSLOCATION",col="black",lty=1,cex=0.75,lwd=1.2,bty='n')

# **Legend** (outer to inner tracks):
# 
# * Ideogram of genome
# * Somatic SNVs: <span style="color:#bd0026"> Potentially actionable </span> <span style="color:#f03b20"> Biologically relevant </span> <span style="color:#fed976"> Variant of Unknown Significance </span>
# * Copy number profiles: <span style="color:red"> Gains (log2FC > 0.6) </span> <span style="color:blue"> Loss (log2FC < 1) </span> <span style="color:grey"> Neutral </span>
# * Structural Variants:
#   - <span style="color:blue"> Deletions, height indicative of VAF </span>
#   - <span style="color:red"> Duplications, height indicative of VAF </span>
#   - <span style="color:green"> Insertions, height indicative of VAF </span>
#   - Lines indicate translocations
#   
# This legend will be replaced with a pictorial guide once the diagram has been finalised.

```

