
#### Overview of germline SV data

Summary of the number of variants:

* ACMG score `r paramfile$AnnotSVACMGcutoff`
* Variant Allele Frequency above `r paramfile$SVVAF`
* Minimum number of supporting split reads `r paramfile$SVSRfilter`
* Minimum number of supporting paired reads `r paramfile$SVPRfilter`

```{r, message=FALSE}
SVTab=read.delim(data$SVSummary, sep="\t")
S2=read.delim(data$SVSplit, sep="\t")
S3=read.delim(data$SVacmg, sep="\t")

## Print the summary table
kable(t(SVTab), rownames = F)
```
The following variants are color coded:

* <span style="color: red;"> Gains or insertions (top) </span>
* <span style="color: blue;"> Deletions (bottom) </span>
* <span style="color: green;"> Inversions and other translocations (top) </span>

Cosmic related genes or genes appearing in user defined list of interest are indicated

```{r, fig.height=5, message=FALSE}

gain=which(S1$TYPE%in%c("DUP", "INS"))
homloss=which(S1$TYPE%in%c("DEL"))
inv=which(S1$TYPE%in%c("INV", "BND"))

gainMat2=tryCatch({CreateGRangesData(S1[gain, ], "Reds",T, 500)}, error=function(e){GRanges()})
lossMat2=tryCatch({CreateGRangesData(S1[homloss, ], "Blues",T, 500)}, error=function(e){ GRanges()})
invMat2=tryCatch({CreateGRangesData(S1[inv, ], "Green",T, 500)}, error=function(e){GRanges()})

if (length(S2$GENE)>1){
  genes=genesAll[na.omit(match(unique(S2$GENE), genesAll$hgnc_symbol))]
}

kp <- plotKaryotype(chromosomes=c("autosomal"), genome="hg38", plot.type=2)
ax=tryCatch({kpPlotRegions(kp, gainMat2, col=gainMat2$AFcol,data.panel = 1)}, error=function(e){NULL})
ax=tryCatch({kpPlotRegions(kp, lossMat2, col=lossMat2$AFcol,data.panel = 2)}, error=function(e){NULL})
ax=tryCatch({kpPlotRegions(kp, invMat2, col=invMat2$AFcol,data.panel = 3)}, error=function(e){NULL})
ax=tryCatch({kpPlotMarkers(kp, data=genes, labels=genes$hgnc_symbol, text.orientation = "horizontal",
              r1=0.75, cex=0.5, adjust.label.position = FALSE)}, error=function(e){NULL})
```
## Genes/Pathways of Interest

```{r}
## Visualise the genes of interest
if (length(S2$MultiSpan)>0){

temp1=S2[which(S2$MultiSpan), c("CHROM", "START", "END","GENE", "Cosmic", "Pathways","TYPE", "Location")]
temp1$Col=ifelse((temp1$Cosmic!=""&temp1$Pathways!=""),"purple", ifelse(temp1$Cosmic!="", "red", "blue") )
temp1$TYPE=factor(temp1$TYPE)

ggplot(temp1, aes(x=GENE, y=1, fill=TYPE))+geom_tile()+theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())+scale_fill_manual(values = fillvals)+expand_limits(y = 0.25)+geom_text(aes(colour = Col, label = GENE), y = 0.35, angle=90)+scale_colour_manual(values = c("blue", "purple", "red"), guide = "none")
}
```

**Multi-gene spanning SVs**

```{r}
idx=which(S1$GenesOfInterest!=""|S1$Cosmic!="")
MGexist=ifelse(length(idx)>0, T, F)
```

`r ifelse(MGexist==F, print('There are no SVs spanning multiple genes'), print('The SVs spanning multiple genes are shown below:'))`

```{r, eval=MGexist}
viewIdx=c("SV_ID", "CHROM", "START","END","SV_length", "BND_partner","GenesOfInterest", "Cosmic", "OMIM_morbid","SplitReadVAF")
HideIdx=setdiff(1:ncol(S1), match(viewIdx, colnames(S1)))-1
DT::datatable(S1[idx, ], rownames=F, filter = 'top', class='cell-border stripe',extensions="Buttons", options=list(dom="Bfrtip",scoller = TRUE, pageLength = 5,  buttons=c(I('colvis'),'csv', 'excel'), columnDefs = list(list(visible=FALSE, targets=HideIdx))))

```

#### Gene level summary

The following fields are gene level annotations from AnnotSV: please see this documentation for more information: https://lbgi.fr/AnnotSV/Documentation/README.AnnotSV_latest.pdf

* GenCC Haploinsufficiency,Triplosensitivity: [0-3] 3 max score for dosage pathogenicity. Scores of 40/30 evidence for no dosage sensitivity and autosomal recessive
* OMIM: disease annotation from OMIM and morbidity from loss
* pLI: intolerance to protein truncating variants from gnomad or ExAC [0 -1]: closer to 1 is predicted intolerance


**COSMIC related genes**

```{r}
viewIdx=c("SV_ID", "GENE", "CHROM", "START","END", "CN", "Location", "Haploinsufficiency", "Triplosensitivity", "OMIM_morbid", "Pathways", "ACMG","Reg.Element.Genes", "Nearest.Splice.Site.bp", "TAD", "BenignSource", "PathogenicSource", "gnomad.pLI","SplitReadVAF", "MeanVAF", "Depth")
HideIdx=setdiff(1:ncol(S2), match(viewIdx, colnames(S2)))-1
CGenes=which(S2$Cosmic!="")
# label parameters to view by default
DT::datatable(S2[CGenes, ], rownames=F, filter = 'top', class='cell-border stripe',extensions="Buttons", options=list(dom="Bfrtip",scoller = TRUE, pageLength = 5,  buttons=c(I('colvis'),'csv', 'excel'), columnDefs = list(list(visible=FALSE, targets=HideIdx))))
```

**Pathway related genes**

```{r}
PGenes=which(S2$GenesOfInterest!="" )
# label parameters to view by default
DT::datatable(S2[PGenes, ], rownames=F, filter = 'top', class='cell-border stripe',extensions="Buttons", options=list(dom="Bfrtip",scoller = TRUE, pageLength = 5,  buttons=c(I('colvis'),'csv', 'excel'), columnDefs = list(list(visible=FALSE, targets=HideIdx))))
```


#### ACMG class 4 or 5 regions

The following variants are level 4 or 5 based on AnnotSV ACMG prediction. Review the genes and population frequencies to assess whether they are relevant or not:

```{r, fig.height=6}
gain=which(S3$SV_type%in%c("DUP", "INS"))
homloss=which(S3$SV_type%in%c("DEL"))
inv=which(S3$SV_type%in%c("INV", "BND"))

gainMat2=tryCatch({CreateGRangesData(S3[gain, ], "Reds",T, 500)}, error=function(e){GRanges()})
lossMat2=tryCatch({CreateGRangesData(S3[homloss, ], "Blues",T, 500)}, error=function(e){ GRanges()})
invMat2=tryCatch({CreateGRangesData(S3[inv, ], "Green",T, 500)}, error=function(e){GRanges()})

GList=setdiff(na.omit(unique(c(S3$GenesOfInterest, S3$Cosmic))), "")

if (length(GList)>0){
  genes=genesAll[na.omit(match(GList, genesAll$hgnc_symbol))]
}
          
kp <- plotKaryotype(chromosomes=c("autosomal"), genome="hg38", plot.type=2)
ax=tryCatch({kpPlotRegions(kp, gainMat2, col=gainMat2$AFcol,data.panel = 1)}, error=function(e){NULL})
ax=tryCatch({kpPlotRegions(kp, lossMat2, col=lossMat2$AFcol,data.panel = 2)}, error=function(e){NULL})
ax=tryCatch({kpPlotRegions(kp, invMat2, col=invMat2$AFcol,data.panel = 3)}, error=function(e){NULL})
ax=tryCatch({kpPlotMarkers(kp, data=genes, labels=genes$hgnc_symbol, text.orientation = "horizontal",
              r1=0.75, cex=0.5, adjust.label.position = FALSE)}, error=function(e){NULL})

```

```{r}
# label parameters to view by default
viewIdx=c("SV_ID","CHROM", "START","END", "GenesOfInterest", "Cosmic", "Haploinsufficiency", "Triplosensitivity", "OMIM_morbid", "ACMG")
HideIdx=setdiff(1:ncol(S3), match(viewIdx, colnames(S3)))-1
DT::datatable(S3, rownames=F, filter = 'top', class='cell-border stripe',extensions="Buttons", options=list(dom="Bfrtip",scoller = TRUE, pageLength = 5,  buttons=c(I('colvis'),'csv', 'excel'), columnDefs = list(list(visible=FALSE, targets=HideIdx))))

```