
```{r}
## Note there is an error if you try to insert gene names

if (!is.null(data$Manta)){

dinsertfilt=dinsert[which(dinsert$ACMG_class>3), ]
dinsertGL=dinsertfilt[which(dinsertfilt$SV_type%in%c("DEL", "DUP", "INS")), ]

bed2=dinsertGL[ c("SV_chrom", "SV_start", "SV_end", "VAFs", "SV_length","SV_type", "chr2", "loc2", "Gene_name")]
bed2[ ,1]=paste("chr", bed2[ ,1], sep="")

}

bed=A1[ ,c("CONTIG","START","END", "MEAN_LOG2_COPY_RATIO")]
colnames(bed)=c("chr","start","end", "value1")
bed[ ,1]=paste("chr", bed[,1], sep="")

bedt=do.call(rbind, Xout[2:4])
bed3=cbind(bedt[ ,c("Chromosome", "Start_Position", "End_Position")], value=substr(rownames(bedt), 1, 1))
bed3$value=gsub("F", 1,bed3$value)
bed3$value=gsub("P", 2,bed3$value)
bed3$value=gsub("V", 3,bed3$value)

bed3[ ,1]=paste("chr", bed3[,1], sep="")
bed3$value=as.numeric(bed3$value)

colnames(bed3)=c("chr","start","end","value1")
## include the bed file for actionable mutations?
## insert gene names of actionable mutations, or those which have a CNV
bed4=bed3[ which(bed3$value!=3), c(1:4) ]



```


```{r}
circos.par("track.height" = 0.08)
circos.initializeWithIdeogram(species="hg19") # plot the ideogram


circos.genomicTrack(bed3, stack=TRUE, panel.fun = function(region,value,...) {
    circos.genomicPoints(region, value, cex = 0.75, pch = 20,col=ifelse(value==1,"#bd0026", ifelse(value==2, "#f03b20", "#fed976")), ...)
})

circos.genomicTrack(bed, panel.fun = function(region, value, ...) {
    circos.genomicPoints(region, value, cex = 0.75, pch = 20,col=ifelse(value>0.6, "red", ifelse(value<(-1), "blue", "grey")), ...)
})

#circos.genomicTrack(bed, panel.fun = function(region, value, ...) {    circos.genomicPoints(region, value, cex = 0.5, pch = 16,col=ifelse(value>0.6, "red", ifelse(value<(-1), "blue", "grey")), ...)})

if (!is.null(data$Manta)){

typeE=c("DEL","DUP","INS")
colE=c("blue","red","green")
for (i in 1:3) {
        bed_list=bed2[which(bed2[,6]==typeE[i]),c(1:4)]
        if (dim(bed_list)[1]>0){
        circos.genomicTrack(bed_list, track.height=0.05,
    panel.fun = function(region, value, ...) {
        circos.genomicRect(region, value, ytop.column = 1, ybottom = 0, 
            col = colE[i], ...)
        circos.lines(CELL_META$cell.xlim, c(0, 0), lty = 2, col = "#00000040")
    })
    } else {
        circos.genomicTrackPlotRegion(bed_list,track.height=0.05,stack=TRUE, panel.fun = function(region, value, ...) {circos.genomicRect(region, value, cex = 0.5, pch = 16, col = colE[i], ...)})
    }

}

## plot the last part of this
bed2b=bed2[which(bed2[ ,6]=="BND"), c(1,2, 7:8)]
if (dim(bed2b)[1]>0){
for (i in 1:dim(bed2b)[1]) {
        circos.link(bed2b[i,1],as.numeric(bed2b[i,2]),bed2b[i,3],as.numeric(bed2b[i,4]))
}}
}

title("Somatic calls (SNV - CNV - SV)")
legend(0.7,1.05,legend=c("SNV", "CNV","SV-DEL","SV-DUP","SV-INS"),col=c("orange","grey","blue","red","green"),pch=c(16,16,15,15,15),cex=0.75,title="Tracks:",bty='n')
legend(0.7,0.8,legend="SV-TRANSLOCATION",col="black",lty=1,cex=0.75,lwd=1.2,bty='n')
```

**Legend** (outer to inner tracks):

* Ideogram of genome
* Somatic SNVs: <span style="color:#bd0026"> Potentially actionable </span> <span style="color:#f03b20"> Biologically relevant </span> <span style="color:#fed976"> Variant of Unknown Significance </span>
* Copy number profiles: <span style="color:red"> Gains (log2FC > 0.6) </span> <span style="color:blue"> Loss (log2FC < 1) </span> <span style="color:grey"> Neutral </span>
* Structural Variants:
  - <span style="color:blue"> Deletions, height indicative of VAF </span>
  - <span style="color:red"> Duplications, height indicative of VAF </span>
  - <span style="color:green"> Insertions, height indicative of VAF </span>
  - Lines indicate translocations
  
This legend will be replaced with a pictorial guide once the diagram has been finalised.